source/language/openpulse.rst

```diff
@@ -0,0 +1,317 @@
+
+The exact precision of these parameters is hardware specific.
+
+frameof command
+~~~~~~~~~~~~~~~
+
+Frames are uniquely identified by a string name and a set of qubits. The order
+of qubits does not matter.
+
+Frames are retrieved using the ``frameof`` function and passing in the name and
+list of physical qubits.
+
+.. code-block: none
+
+   frameof("drive", %0)
+   
+   // These next two lines refer to the same frame
+   frameof("coupling", %0, %1)
+   frameof("coupling", %1, %0)
+
+Frame names may seem like they ascribe meaning or that there are only certain
+permissable names. This is not the case; frame names are arbitrary. The frames
+are later mapped to channel names that do have meaning for a certain hardware
+vendor. For example, the hardware vendor may choose to map frames to channels
+using JSON:
+
+.. code-block: javascript
+
+   {
+     drive: {
+       "{0}": "channel0",
+       "{1}": "channel1"
+     coupling: {
+       "{0,1}": "channel2"
+     }
+   }
+
+This has the advantage that one can run any program with any arbitrary frame
+names provided a mapping to the appropriate channels is given.
+
+Restrictions on the use of frames
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+There are two important considersations when dealing with frames.
+
+First, frames are compile time constants. Runtime variables cannot be used as
+arguments to frames; the frame must be resolvable at compile time. This also
+means that assigning a frame to an alias (for the purpose of avoiding typing)
+needs to be done with ``const`` not ``let``.
+
+.. code-block: none
+
+   const driveframe = frameof("drive", %0);
+
+Second, frames return references not values. This means that calling the same
+``frameof`` command in different places is the same as assigning the frame to a
+constant and referencing it multiple times.
+
+Frame manipulation
+~~~~~~~~~~~~~~~~~~
+
+The (frequency, phase, scale) of a frame can be manipulated throughout program
+by referencing ``.frequency``, ``.phase``, and ``.scale``. Operations must be 
+appropriate for the respective type, ``float`` for frequency, ``angle`` for
+phase, and ``float`` for scale. Again, the exact precision of these calculations
+is hardware specific.
+
+Here's an example of manipulating the phase to calibrate an ``rz`` gate:
+
+.. code-block:: none
+
+   // Shift phase of qubit 0's "drive" frame by pi/4, eg. an rz gate with angle -pi/4
+   frameof("drive", %0).phase += pi/4;
+
+   // Equivalent
+   const drive = frameof("drive", %0);
+   drive.phase += pi/4;
+
+   // Define a calibration for the rz gate on all physical qubits
+   defcal rz(angle[20]:theta) %q {
+     frameof("drive", %q).phase -= theta;
+   }
+
+Here's an example qubit spectroscopy experiment.
+
+.. code-block:: none
+
+   qubit q;
+
+   const shots = 1000;
+   const start = 5e9; // 5 GHz
+   const end = 6e9; // 6 GHz
+   const points = 50;
+
+   complex[32] iq, average;
```

Comment:
I don't think `complex` types have yet been defined in OpenQASM. Probably a good idea.